File: library/src/main/java/no/wtw/android/restserviceutils/UserAgentFormatter.java
package no.wtw.android.restserviceutils;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.DisplayMetrics;

public class UserAgentFormatter {

    public static String getUserAgent(Context context) {
        return getAppName(context) + "/" + getVersion(context) + " " + getDevice() + " (" + getScreenSize(context) + ") Android/" + Build.VERSION.RELEASE;
    }

    private static String getAppName(Context context) {
        return context.getApplicationInfo().loadLabel(context.getPackageManager()).toString();
    }

    public static String getVersion(Context context) {
        PackageManager manager = context.getPackageManager();
        try {
            PackageInfo info = manager.getPackageInfo(context.getPackageName(), 0);
            return info.versionName + " (" + info.versionCode + ")";
        } catch (PackageManager.NameNotFoundException e) {
            return "?";
        }
    }

    public static String getScreenSize(Context context) {
        DisplayMetrics metrics = context.getResources().getDisplayMetrics();
        return metrics.widthPixels + "x" + metrics.heightPixels + "@" + metrics.density;
    }

    public static String getDevice() {
        return Build.MODEL + "/" + Build.DEVICE;
    }
}

File: library/src/main/java/no/wtw/android/restserviceutils/UserAgentFormatter.kt
package no.wtw.android.restserviceutils

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build

open class UserAgentFormatter(
    private val context: Context,
) {

    val userAgent: String
        get() = appName + "/" + version + " " + device + " (" + screenSize + ") Android/" + Build.VERSION.RELEASE

    val appName: String
        get() = context.applicationInfo.loadLabel(context.packageManager).toString()

    val version: String
        get() = try {
            context.packageManager.getPackageInfo(context.packageName, 0).let { it.versionName + " (" + it.versionCode + ")" }
        } catch (e: PackageManager.NameNotFoundException) {
            "?"
        }

    val screenSize: String
        get() = context.resources.displayMetrics.let { it.widthPixels.toString() + "x" + it.heightPixels + "@" + it.density }

    val device: String
        get() = Build.MODEL + "/" + Build.DEVICE

}